### PR TITLE
[TESTERS NEEDED] LLVM: New pass manager

### PR DESCRIPTION
**Pull request by oltolm**
RPCS3 was using the old pass manager and I made it use the new one. This would allows us to switch to LLVM 17, but it also works with LLVM 16.

### CODE REVIEW

**Comment by elad335** [comment]
Needs testing for performance regressions

**Comment by Ordinary205** [comment]
I can confirm that I'm experiencing performance regression while I was testing it on Demon's Souls and Red Dead Redemption.
Demon's Souls:
Master: 86.9/91/95 FPS
Demon's Souls master

PR: 85.4/89.7/92.2 FPS
Demon's Souls pull request

Red Dead Redemption:
Master: 29.8/30.8/31.7 FPS
RDR master

PR: 28.2/29.1/30 FPS
RDR pull request

[DS.log.gz](https://github.com/RPCS3/rpcs3/files/13687673/DS.log.gz)
[RDR.log.gz](https://github.com/RPCS3/rpcs3/files/13687675/RDR.log.gz)

**Comment by oltolm** [comment]
Thanks for testing. Does anyone know what the error margin is?

**Comment by oltolm** [comment]
Please retest.

**Comment by Ordinary205** [comment]
Demon's Souls seems to have slightly better performance than master, but sadly Red Dead Redemption still has low performance regression.
Demon's Souls:
Master: 86.9/91/95 FPS
PR: 86.7/93.1/96.2 FPS
Demon's Souls pull request 2

Red Dead Redemption:
Master: 29.8/30.8/31.7 FPS
PR: 27.7/29.2/30 FPS
RDR pull request 2

[DS.log.gz](https://github.com/RPCS3/rpcs3/files/13692191/DS.log.gz)
[RDR.log.gz](https://github.com/RPCS3/rpcs3/files/13692192/RDR.log.gz)

**Comment by oltolm** [comment]
I don't know why RDR2 is slower. Maybe someone wants to find out and experiment with different combinations of [LLVM compiler passes](https://llvm.org/docs/Passes.html). See also the documentation of the [new pass manager](https://llvm.org/docs/NewPassManager.html).

**Comment by elad335** [comment]
Needs retesting performance with games.

**Comment by oltolm** [comment]
I have changed the code to use the old pass manager for LLVM < 17 and the new one for LLVM >= 17. We could therefore merge it as is. The new PM won't be used for the official build.

File: rpcs3/Emu/Cell/SPULLVMRecompiler.cpp [hunk 1]
```diff
@@ -2507,24 +2546,9 @@ class spu_llvm_recompiler : public spu_recompiler_base, public cpu_translator
...
-		legacy::FunctionPassManager pm(_module.get());
-
-		// Basic optimizations
-		pm.add(createEarlyCSEPass());
-		pm.add(createCFGSimplificationPass());
-#if LLVM_VERSION_MAJOR < 17
-		pm.add(createDeadStoreEliminationPass());
-		pm.add(createAggressiveDCEPass());
-#else
-		pm.add(createDeadCodeEliminationPass());
```
Comment (elad335) [inline rpcs3/Emu/Cell/SPULLVMRecompiler.cpp:2520]: No replacement here? even if unused.

**Comment by Ordinary205** [comment]
Red Dead Redemption seems to be running better than before.
Master: 25.8/27.2/28.3 FPS
<details>
<summary>Screenshot</summary>

Red Dead Redemption Before

</details>

PR: 27.2/28.4/29.3 FPS
<details>
<summary>Screenshot</summary>

Red Dead Redemption After

</details>

[Master.log.gz](https://github.com/RPCS3/rpcs3/files/14808285/Master.log.gz)
[PR.log.gz](https://github.com/RPCS3/rpcs3/files/14808286/PR.log.gz)
So merging this PR should be fine without having to worry about performance downgrade.

**Comment by oltolm** [comment]
It runs the same as before because the new PM is not used when compiled against LLVM 16 and this build is compiled against LLVM 16.

**Pull request merged**